ci/Debian11Container.Jenkinsfile

```diff
@@ -0,0 +1,303 @@
+						unstash 'native_modules'
+
+						// add DEBUG for electron-builder because it tends to not let us know about things failing
+						withCredentials([string(credentialsId: 'HSM_USER_PIN', variable: 'PW')]) {
+							sh '''
+							export HSM_USER_PIN=${PW};
+							export WIN_CSC_FILE="/opt/etc/codesign.crt";
+							DEBUG=electron-builder node desktop --existing --platform win '''
+						}
+
+						dir('artifacts') {
+							stash includes: 'desktop-test/*', name:'win_installer_test'
+							stash includes: 'desktop/*', name:'win_installer'
+						}
+					}
+                }
+                stage('Mac') {
+					environment {
+						PATH = "${env.NODE_MAC_PATH}:${env.PATH}"
+					}
+					agent {
+						label 'mac-m1'
+					}
+					steps {
+						initBuildArea()
+
+						withCredentials([
+								usernamePassword(credentialsId: 'APP_NOTARIZE_CREDS', usernameVariable: 'APPLEIDVAR', passwordVariable: 'APPLEIDPASSVAR'),
+								string(credentialsId: 'fastlane-keychain-password', variable: 'FASTLANE_KEYCHAIN_PASSWORD'),
+								string(credentialsId: 'team-id', variable: 'APPLETEAMIDVAR'),
+						]) {
+							sh 'security unlock-keychain -p $FASTLANE_KEYCHAIN_PASSWORD'
+							script {
+								def stage = params.RELEASE ? 'release' : 'prod'
+								sh '''
+								export APPLEID=${APPLEIDVAR};
+								export APPLEIDPASS=${APPLEIDPASSVAR};
+								export APPLETEAMID=${APPLETEAMIDVAR};
+								node desktop --existing --architecture universal --platform mac ''' + "${stage}"
+								dir('artifacts') {
+									if (params.RELEASE) {
+										stash includes: 'desktop-test/*', name:'mac_installer_test'
+									}
+									stash includes: 'desktop/*', name:'mac_installer'
+								}
+							}
+						}
+					}
+				}
+				stage('Linux') {
+					agent {
+						dockerfile {
+							filename 'Debian11Container.dockerfile'
+							label 'master'
+							dir 'ci'
+							additionalBuildArgs "--format docker"
+							args '--network host'
+						} // docker
+					}
+					steps {
+						initBuildArea()
+
+						sh 'node desktop --existing --platform linux'
+
+						dir('artifacts') {
+							stash includes: 'desktop-test/*', name:'linux_installer_test'
+							stash includes: 'desktop/*', name:'linux_installer'
+						}
+					}
+				}
+			}
+		}
+
+		stage('Preparation for build deb and publish') {
+			when { expression { params.RELEASE } }
+			agent {
+				label 'linux'
+			}
+			steps {
+				script {
+					def devicePath =  sh(script: 'lsusb | grep Nitro | sed -nr \'s|Bus (.*) Device ([^:]*):.*|/dev/bus/usb/\\1/\\2|p\'', returnStdout: true).trim()
+					env.DEVICE_PATH = devicePath
+				}
+			}
+		}
+		stage('Build deb and publish') {
+			when { expression { params.RELEASE } }
+			agent {
+				dockerfile {
+					filename 'Debian11Container.dockerfile'
+					label 'master'
+					dir 'ci'
+					additionalBuildArgs '--format docker'
+					args "--network host -v /run:/run:rw,z -v /opt/repository:/opt/repository:rw,z --device=${env.DEVICE_PATH}"
```

Comment:
when looking at the additional step for getting the device I had a thought of whether it would be possible to make a function and not a separate step to fetch it so that we can write `--device=${getHsmDevicePath()}` but I'm not sure if it works like that in Jenkins.

If we just reference environment variable then perhaps the single quotes would be a bit better? see https://docs.cloudbees.com/docs/cloudbees-ci/latest/automating-with-jenkinsfile/string-interpolation

Reply:
I will just mention here for documentation purposes. We tried this together and it did not work due to `java.lang.NoSuchMethodError: No such DSL method 'getDevicePath' found among steps`. We decided to do it when switching to imperative pipelines.